### DISCUSSION TITLE
Avoiding local optimum

### DISCUSSION THREAD

**Discussion by A-ep93** [meta-pytorch/botorch#1182] [Q&A]
Hello,
I am using botorch for a problem and it works great most of the time. The only problem is that sometimes when I run the optimization, it does not find the global optimum. Then, when I run the optimization for the exact same problem again, the optimizer gives me the global optimum. It seems that for these cases, the model sometimes sticks in a local optimum rather than the global one. I have attached the main parts of my code here and I appreciate it if you can help me with some suggestions.

```
tkwargs = {
    "dtype": torch.double,
    "device": torch.device("cuda" if torch.cuda.is_available() else "cpu"),
}


def problem(input):
        # This is the function I want to optimize
        return torch.tensor(result,**tkwargs)

bounds=torch.tensor([[0],[1]],**tkwargs)

def generate_initial_data(n=6):
        # generate training data
        train_x = draw_sobol_samples(
            bounds=bounds,n=1, q=n, seed=torch.randint(1000000, (1,)).item()
        ).squeeze(0)
        train_obj = problem(train_x)
        best=train_obj.max()
        return train_x, train_obj, best


def get_next_points(train_x, train_obj, best, bounds, n_points):
        models =SingleTaskGP(train_x,train_obj, outcome_transform=Standardize(m=1))
        mll = ExactMarginalLogLikelihood(models.likelihood, models)
        fit_gpytorch_model(mll)
        EI=ExpectedImprovement(
            model=models,
            best_f=best)
        candidate, _ =optimize_acqf(
            acq_function=EI,
            bounds=bounds,
            q=n_points,
            num_restarts=20,
            raw_samples=512,
            )
        return candidate


n_runs=500
train_x, train_obj, best=generate_initial_data(50)



for i in range(n_runs):
        new_candidate=get_next_points(train_x, train_obj, best, bounds, 1)
        new_results=problem(new_candidate)
        train_x=torch.cat([train_x,new_candidate])
        train_obj=torch.cat([train_obj,new_results])
        best=train_obj.max()
        location=train_x[torch.argmax(train_obj)]

```

**Comment by Balandat** [comment]
What dimension is your problem? And how many evaluations does it typically take to get near the optimum if that is successful? Trying to understand whether this behavior is just an unlucky initialization or whether the optimization indeed does get stuck and wouldn't explore other regions of the domain. Have you looked at the model fits / do you have a sense for how good they are?

**Reply by A-ep93** [reply]
Thanks for your reply @Balandat. The dimension of my problem is 2. Based on my observations, usually about 300 iterations in enough for finding the correct global optimum. I am also not sure whether this problem is related to an unlucky initialization or the model gets stuck in a local optimum. I guess maybe the latter one is more probable. 
Regarding the model fits, no I have not checked that. I appreciate it if you can inform me how can I check this in botorch and also I appreciate it if you can tell me other suggestions that comes to your mind.

**Reply by A-ep93** [reply]
Any comments/suggestions regarding this issue are greatly appreciated.

**Reply by Balandat** [reply]
This tutorial shows how to perform cross-validation of the model: https://botorch.org/tutorials/batch_mode_cross_validation - You should be able to adapt this to your data. 

When you say the optimizer finds a local optimum, how much worse is this than the global one? If there is not a large difference in objective values between those (relative to the range of the  function value across the whole search space), it might be hard for Bayesian Optimization to discern between the two.

Also, it would be good to understand how smooth the response surface is of this function you're trying to optimize, and whether it's well suited to be modeled by a Gaussian Process. Since it's 2d it should be pretty straight forward to visualize this with heatmap or contour plot .
